File: DataInjectorIOS/DataInjector/Classes/injector/LinkDataInjector.swift
//
//  LinkDataInjector.swift
//  DataInjector Pod
//
//  Library injector: link datasets
//  Inject one data set into another by a common field that links them together (such as ID)
//

import Foundation

/// An injector linking multiple datasets together by a common field (like ID)
open class LinkDataInjector: BaseInjector {
    
    // --
    // MARK: Members
    // --

    public var targetItemPath: InjectorPath?
    public var sourceDataPath: InjectorPath?
    public var overrideSourceData: Any?
    public var linkKey = "unknown"

    
    // --
    // MARK: Initialization
    // --
    
    public override init() {
    }

    
    // --
    // MARK: Manual injection
    // --
    
    public static func linkData(inDictionary: [String: Any?], fromArray: [Any?], usingKey: String) -> InjectorResult {
        if let foundItem = findDataItem(inArray: fromArray, forValue: inDictionary[usingKey] ?? nil, usingKey: usingKey) {
            var modifiedDictionary = [String: Any?]()
            for (key, value) in inDictionary {
                modifiedDictionary[key] = value
            }
            for (key, value) in foundItem {
                if key != usingKey {
                    modifiedDictionary[key] = value
                }
            }
            return InjectorResult(withModifiedObject: modifiedDictionary)
        }
        return InjectorResult(withModifiedObject: inDictionary)
    }
    
    public static func linkData(onArray: [Any?], fromArray: [Any?], usingKey: String) -> InjectorResult {
        var modifiedData = [[String: Any?]?]()
        for arrayItem in onArray {
            if let dictItem = arrayItem as? [String: Any?] {
                let result = LinkDataInjector.linkData(inDictionary: dictItem, fromArray: fromArray, usingKey: usingKey)
                if result.hasError() {
                    return result
                }
                modifiedData.append(result.modifiedObject as? [String: Any?])
            } else {
                return InjectorResult(withError: .targetInvalid)
            }
        }
        return InjectorResult(withModifiedObject: modifiedData)
    }
    

    // --
    // MARK: Data helpers
    // --
    
    public static func findDataItem(inArray: [Any?], forValue: Any?, usingKey: String) -> [String: Any?]? {
        if let searchValueString = InjectorConv.asString(value: forValue) {
            for arrayItem in inArray {
                if let dictItem = arrayItem as? [String: Any?] {
                    if let compareValueString = InjectorConv.asString(value: dictItem[usingKey] ?? nil) {
                        if compareValueString == searchValueString {
                            return dictItem
                        }
                    }
                }
            }
        }
        return nil
    }
    
    
    // --
    // MARK: General injection
    // --
    
    open override func appliedInjection(targetData: Any?, sourceData: Any? = nil) -> InjectorResult {
        var useSourceData = overrideSourceData ?? sourceData
        useSourceData = DataInjector.get(from: useSourceData, path: sourceDataPath ?? InjectorPath(path: ""))
        return DataInjector.inject(into: targetData, path: targetItemPath ?? InjectorPath(path: ""), modifyCallback: { originalData in
            if let sourceArray = useSourceData as? [Any?] {
                if let targetDict = originalData as? [String: Any?] {
                    return LinkDataInjector.linkData(inDictionary: targetDict, fromArray: sourceArray, usingKey: self.linkKey)
                } else if let targetArray = originalData as? [Any?] {
                    return LinkDataInjector.linkData(onArray: targetArray, fromArray: sourceArray, usingKey: self.linkKey)
                } else {
                    return InjectorResult(withError: .targetInvalid)
                }
            }
            return InjectorResult(withError: .sourceInvalid)
        })
    }

}

File: DataInjectorAndroid/DataInjectorLib/src/main/java/com/crescentflare/datainjector/injector/LinkDataInjector.java
package com.crescentflare.datainjector.injector;

import com.crescentflare.datainjector.conversion.InjectorConv;
import com.crescentflare.datainjector.transformer.BaseTransformer;
import com.crescentflare.datainjector.utility.InjectorPath;
import com.crescentflare.datainjector.utility.InjectorResult;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Data injector: link datasets
 * Inject one data set into another by a common field that links them together (such as ID)
 */
public class LinkDataInjector extends BaseInjector
{
    // --
    // Members
    // --

    private List<? extends BaseTransformer> sourceTransformers = new ArrayList<>();
    private InjectorPath targetItemPath;
    private InjectorPath sourceDataPath;
    private Object overrideSourceData;
    private String linkKey = "unknown";


    // --
    // Initialization
    // --

    public LinkDataInjector()
    {
    }


    // --
    // Manual injection
    // --

    @NotNull
    public static InjectorResult linkData(@NotNull Map<String, Object> targetMap, @NotNull List<?> sourceData, @NotNull String key)
    {
        Map<String, Object> foundItem = findDataItem(sourceData, targetMap.get(key), key);
        if (foundItem != null)
        {
            HashMap<String, Object> modifiedMap = new HashMap<>();
            for (String setKey : targetMap.keySet())
            {
                modifiedMap.put(setKey, targetMap.get(setKey));
            }
            for (String setKey : foundItem.keySet())
            {
                if (!setKey.equals(key))
                {
                    modifiedMap.put(setKey, foundItem.get(setKey));
                }
            }
            return InjectorResult.withModifiedObject(modifiedMap);
        }
        return InjectorResult.withModifiedObject(targetMap);
    }

    @NotNull
    public static InjectorResult linkData(@NotNull List<?> targetList, @NotNull List<?> sourceData, @NotNull String key)
    {
        List<Map<String, Object>> modifiedData = new ArrayList<>();
        for (Object listItem : targetList)
        {
            Map<String, Object> mapItem = InjectorConv.asStringObjectMap(listItem);
            if (mapItem != null)
            {
                InjectorResult result = LinkDataInjector.linkData(mapItem, sourceData, key);
                if (result.hasError())
                {
                    return result;
                }
                modifiedData.add(InjectorConv.asStringObjectMap(result.getModifiedObject()));
            }
            else
            {
                return InjectorResult.withError(InjectorResult.Error.TargetInvalid);
            }
        }
        return InjectorResult.withModifiedObject(modifiedData);
    }


    // --
    // Data helpers
    // --

    @Nullable
    public static Map<String, Object> findDataItem(@NotNull List<?> list, @Nullable Object value, @NotNull String key)
    {
        String searchValueString = InjectorConv.asString(value);
        if (searchValueString != null)
        {
            for (Object listItem : list)
            {
                Map<String, Object> mapItem = InjectorConv.asStringObjectMap(listItem);
                if (mapItem != null)
                {
                    String compareValueString = InjectorConv.asString(mapItem.get(key));
                    if (compareValueString != null)
                    {
                        if (compareValueString.equals(searchValueString))
                        {
                            return mapItem;
                        }
                    }
                }
            }
        }
        return null;
    }


    // --
    // General injection
    // --

    @Override
    @NotNull
    protected InjectorResult onApply(@Nullable Object targetData, @Nullable Object sourceData)
    {
        // Prepare source data with optional transformation
        Object checkSourceData = overrideSourceData != null ? overrideSourceData : sourceData;
        checkSourceData = DataInjector.get(checkSourceData, sourceDataPath != null ? sourceDataPath : new InjectorPath());
        for (BaseTransformer transformer : sourceTransformers)
        {
            InjectorResult result = transformer.apply(checkSourceData);
            if (result.hasError())
            {
                return result;
            }
            checkSourceData = result.getModifiedObject();
        }

        // Apply injection
        final Object useSourceData = checkSourceData;
        return DataInjector.inject(targetData, targetItemPath != null ? targetItemPath : new InjectorPath(), new DataInjector.ModifyCallback()
        {
            @Override
            public @NotNull InjectorResult modify(@Nullable Object originalData)
            {
                List<Object> sourceList = InjectorConv.asObjectList(useSourceData);
                if (sourceList != null)
                {
                    Map<String, Object> targetMap = InjectorConv.asStringObjectMap(originalData);
                    List<Object> targetList = InjectorConv.asObjectList(originalData);
                    if (targetMap != null)
                    {
                        return LinkDataInjector.linkData(targetMap, sourceList, linkKey);
                    }
                    else if (targetList != null)
                    {
                        return LinkDataInjector.linkData(targetList, sourceList, linkKey);
                    }
                    else
                    {
                        return InjectorResult.withError(InjectorResult.Error.TargetInvalid);
                    }
                }
                return InjectorResult.withError(InjectorResult.Error.SourceInvalid);
            }
        });
    }


    // --
    // Set values
    // --

    public void setSourceTransformers(@Nullable List<? extends BaseTransformer> sourceTransformers)
    {
        this.sourceTransformers = sourceTransformers != null ? sourceTransformers : new ArrayList<BaseTransformer>();
    }

    public void setTargetItemPath(@Nullable InjectorPath targetItemPath)
    {
        this.targetItemPath = targetItemPath;
    }

    public void setSourceDataPath(@Nullable InjectorPath sourceDataPath)
    {
        this.sourceDataPath = sourceDataPath;
    }

    public void setOverrideSourceData(@Nullable Object overrideSourceData)
    {
        this.overrideSourceData = overrideSourceData;
    }

    public void setLinkKey(@Nullable String linkKey)
    {
        this.linkKey = linkKey != null ? linkKey : "unknown";
    }
}
